### PR TITLE
feat(evaluator): Add support for global license choices

### LINKED ISSUE #4379: Curation.yml Configuration Capability Enhancement Request
**Description:** 
During the analysis, we found there are multiple ways of curating license(s) using either curation.yml or .ort.yml. Can "License Finding", "Policy Rule Violation" and "License Choices" also be part of curation.yml as well?

**Reason:** 
As the developers are not aware of the license model(s) and what to curate, this is a dedicated responsibility of curation team in conjunction with Legal team which will allow curation team to configure the license finding using Curation.yml

**Understanding:**
If there is already a way to conclude multiple licenses in Curation.yml, then we would like to understand how?

**Proposal:**
curation-config

### PR DESCRIPTION
**Pull request by tsteenbe**
Add support for defining global license choices in a `license-choices.yml` file in the ORT configuration directory, with the same API as the `.ort.yml`. If the same license choice exists in the `.ort.yml`, it takes precedence over the global license choice.

## Example license-choices.yml file format
```
---
package_license_choices:
- package_id: "PyPI::six:1.17.0"
  license_choices:
  - given: "LicenseRef-ort-example-license OR MIT"
    choice: "MIT"
repository_license_choices:
- given: "LicenseRef-ort-example-license2 OR MIT"
  choice: "MIT"
- given: "LicenseRef-ort-example-license3 OR MIT"
  choice: "MIT"
```

Closes #4379.

### CODE REVIEW

File: plugins/commands/evaluator/src/main/kotlin/EvaluateCommand.kt [hunk 1]
```diff
@@ -278,6 +289,14 @@ class EvaluateCommand(descriptor: PluginDescriptor = EvaluateCommandFactory.desc
             ortResultInput = ortResultInput.replaceConfig(config)
         }
 
+        licenseChoicesFile.takeIf { it.isFile }?.let {
+            val globalLicenseChoices = it.readValue<LicenseChoices>()
+            val config = ortResultInput.repository.config
+            ortResultInput = ortResultInput.replaceConfig(
+                config.copy(licenseChoices = config.licenseChoices.merge(globalLicenseChoices))
```
Comment (fviernau) [inline plugins/commands/evaluator/src/main/kotlin/EvaluateCommand.kt]: Hm, simply adding the choices into the RepositoryConfiguration makes the choices be applied but can have problems

1. One would assume all entries come from `.ort.yml`
2. If one write a rule to check .ort.yml validity, such rule could no more distinguish and would run over global entries too.

Looking how it was done for other configration items, I believe the global license choices can be done similar to how it is already implemented for resolutions. I don't see a need for introducing a provider, but the choices could be added to `ResolvedConfiguration` too. 

Note that in the reporter,command there is a corresponding `--refresh` option, which here probably should be done consistently too.

**Review by fviernau — CHANGES_REQUESTED** [review]

File: plugins/commands/evaluator/src/main/kotlin/EvaluateCommand.kt (hunk 1 above)
Comment (tsteenbe) [reply plugins/commands/evaluator/src/main/kotlin/EvaluateCommand.kt]: @fviernau I just pushed a new version that implements things per our last week's session. Would you mind doing another review?

File: model/src/main/kotlin/utils/LicenseChoiceUtils.kt [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+/** License views to compute effective licenses for, to use for filtering license choices. */
+private val FILTER_LICENSE_VIEWS = setOf(
+    LicenseView.ONLY_DECLARED,
+    LicenseView.ONLY_DETECTED,
+    LicenseView.CONCLUDED_OR_DECLARED_AND_DETECTED
+)
+
+fun LicenseChoices.filterApplicable(ortResult: OrtResult): LicenseChoices {
+    val packageIds = ortResult.getPackages().mapTo(mutableSetOf()) { it.metadata.id }
```
Comment (fviernau) [inline model/src/main/kotlin/utils/LicenseChoiceUtils.kt:40]: Just to double check, license choices get really applied only to packages, not to projects?

File: model/src/test/kotlin/config/LicenseChoicesTest.kt [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            val global = LicenseChoices(
+                packageLicenseChoices = listOf(
+                    PackageLicenseChoice(
+                        id,
+                        listOf(
+                            SpdxLicenseChoice("A OR B".toSpdx(), "B".toSpdx()),
+                            SpdxLicenseChoice("E OR F".toSpdx(), "E".toSpdx())
+                        )
+                    ),
+                    PackageLicenseChoice(otherId, listOf(SpdxLicenseChoice("G OR H".toSpdx(), "G".toSpdx())))
```
Comment (fviernau) [inline model/src/test/kotlin/config/LicenseChoicesTest.kt:67]: I believe the choices defined in lines 65 and 68 can be removed to make the test more focused on a single thing, which IIUC is the way things are overridden. 

Aprt from that, for package license chocies, it is possible the `given` is `null`. Should we test this behavior too?

File: model/src/test/kotlin/config/LicenseChoicesTest.kt [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+            val merged = repository.merge(global)
+
+            merged.repositoryLicenseChoices shouldContainExactly listOf(
+                SpdxLicenseChoice("A OR B".toSpdx(), "A".toSpdx()),
+                SpdxLicenseChoice("C OR D".toSpdx(), "C".toSpdx())
+            )
+        }
+
+        "let repository package choices take precedence over global package choices" {
```
Comment (fviernau) [inline model/src/test/kotlin/config/LicenseChoicesTest.kt]: Remark: The `merge()` function has a bit of edge cases to be tested. 
Have you considered that it is possible, remove almost all the logic and simply 
append this lists for choices for a particular `id` ?

As the chocies get applied in order of the lists, having global choices come last also should lead to the desired functionality.

File: model/src/main/kotlin/config/LicenseChoices.kt [hunk 5]
```diff
@@ -45,6 +45,31 @@ data class LicenseChoices(
...
+     */
+    fun merge(other: LicenseChoices): LicenseChoices {
+        val givens = repositoryLicenseChoices.mapTo(mutableSetOf()) { it.given }
+        val mergedRepositoryLicenseChoices = repositoryLicenseChoices +
+            other.repositoryLicenseChoices.filterNot { it.given in givens }
+
+        val packageIds = (packageLicenseChoices.map { it.packageId } + other.packageLicenseChoices.map { it.packageId })
+            .distinct()
+        val mergedPackageLicenseChoices = packageIds.map { packageId ->
+            val choices = packageLicenseChoices.find { it.packageId == packageId }?.licenseChoices.orEmpty()
```
Comment (fviernau) [inline model/src/main/kotlin/config/LicenseChoices.kt]: This `find()` function only finds the first  occurrence. Could it be that all further occurences are missed?

Note: It can be that each, global and local package license choices, contain multiple entries for a particular id. 

I propose for now to leave it like (in this PR), and not start merging entries for a particular id. Because 
1. then choices entries remain original and are thus easier to debug
2. the logic remain very simple (contat two lists with the plus operator.

Bottom line: what do you think about re-writing merge as follows?

```
LicenseChoices(
   packageLicensesChoices = packageLicensesChoices + other.packageLicensesChoices,
   repositoryLicensesChoices = repositoryLicensesChoices + other.repositoryLicensesChoices
)
```

**Review by fviernau — CHANGES_REQUESTED** [review]

File: integrations/completions/ort-completion.bash [hunk 6]
```diff
@@ -787,6 +787,12 @@ _ort_evaluate() {
           [[ ${i} -gt COMP_CWORD ]] && in_param='--copyright-garbage-file' || in_param=''
           continue
           ;;
+        --license-choices-file)
```
Comment (fviernau) [inline integrations/completions/ort-completion.bash:790]: I'd tend to squash the topmost 2 commits as implementing this only for the evaluator will lead 
to choices not working if evaluator is not used. Having it squashed, everything should remain working.

File: integrations/completions/ort-completion.bash [hunk 7]
```diff
@@ -787,6 +787,12 @@ _ort_evaluate() {
...
+          __skip_opt_eq
+          (( i = i + 1 ))
+          [[ ${i} -gt COMP_CWORD ]] && in_param='--license-choices-file' || in_param=''
```
Comment (fviernau) [inline integrations/completions/ort-completion.bash:793]: commit-msg: We normally uses "Fixes" instead of "Closes" IIRC.

**Review by fviernau — CHANGES_REQUESTED** [review]

File: plugins/commands/reporter/src/main/kotlin/ReportCommand.kt [hunk 8]
```diff
@@ -245,7 +258,16 @@ class ReportCommand(descriptor: PluginDescriptor = ReportCommandFactory.descript
         }
 
         ortResult = ortResult.setPackageConfigurations(packageConfigurationProvider)
-            .setLicenseChoices(ortResult.repository.config.licenseChoices)
+
+        val applicableGlobalLicenseChoices = licenseChoicesFile
+            .takeIf { it.isFile }
+            ?.let { it.readValue<LicenseChoices>() }
```
Comment (fviernau) [inline plugins/commands/reporter/src/main/kotlin/ReportCommand.kt]: This line should be changed to just `?.readValue<LicenseChoices>()`.
Same in the evaluator command of the other commit.

File: plugins/commands/reporter/src/test/resources/test-ort-result.yml [hunk 9]
```diff
@@ -0,0 +1,13 @@
+---
```
Comment (fviernau) [inline plugins/commands/reporter/src/test/resources/test-ort-result.yml]: This is unused, or?

**Review by fviernau — CHANGES_REQUESTED** [review]

**Review by fviernau — DISMISSED** [review]

File: evaluator/src/testFixtures/kotlin/TestData.kt [hunk 10]
```diff
@@ -188,24 +204,12 @@ val ortResult = OrtResult(
...
-                        )
-                    )
-                )
-            )
+            licenseChoices = licenseChoices
         )
     ),
+    // Mirror the production flow where repository (and global) license choices are resolved into the resolved
+    // configuration, which is where `getRepositoryLicenseChoices()` reads them from.
+    resolvedConfiguration = ResolvedConfiguration(licenseChoices = licenseChoices),
```
Comment (fviernau) [inline evaluator/src/testFixtures/kotlin/TestData.kt:210]: Should the diff in this file go into the previous commit?

Comment (tsteenbe) [reply evaluator/src/testFixtures/kotlin/TestData.kt:210]: Yes, you are right fixed now.

**Review by fviernau — APPROVED** [review]
